File: src/main/java/dregsmod/actions/SealCardAction.java
package dregsmod.actions;

import com.megacrit.cardcrawl.actions.AbstractGameAction;
import com.megacrit.cardcrawl.actions.GameActionManager;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.CardGroup;
import com.megacrit.cardcrawl.core.Settings;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import dregsmod.patches.variables.CardSealed;
import dregsmod.vfx.SealCardEffect;

import java.util.logging.Logger;

public class SealCardAction extends AbstractGameAction {
    private static final Logger logger = Logger.getLogger(SealCardAction.class.getName());
    private AbstractCard card;
    private CardGroup group;

    public SealCardAction(AbstractCard card, CardGroup group) {
        duration = Settings.ACTION_DUR_FASTER;
        setValues(AbstractDungeon.player, AbstractDungeon.player, 1);
        actionType = ActionType.DISCARD;
        this.card = card;
        this.group = group;
    }

    public SealCardAction(AbstractCard card) {
        this(card, AbstractDungeon.player.hand);
    }

    @Override
    public void update() {
        isDone = true;
        if (group.contains(card)) {
            AbstractDungeon.effectList.add(new SealCardEffect(card));

            CardSealed.isSealed.set(card, true);
            card.exhaustOnUseOnce = false;
            card.freeToPlayOnce = false;

            group.moveToDiscardPile(card);
            GameActionManager.incrementDiscard(false);

            card.triggerOnExhaust();
            card.triggerOnManualDiscard();
        }

        tickDuration();
    }
}

File: src/main/java/accursed/actions/SealCardAction.kt
package accursed.actions

import accursed.AccursedMod
import com.megacrit.cardcrawl.actions.AbstractGameAction
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.cards.CardGroup
import com.megacrit.cardcrawl.core.Settings
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.vfx.cardManip.ExhaustCardEffect

class SealCardAction(val card: AbstractCard, private val group: CardGroup) : AbstractGameAction() {

    init {
        duration = Settings.ACTION_DUR_FAST
        setValues(AbstractDungeon.player, AbstractDungeon.player, 1)
    }

    override fun update() {
        if (duration == Settings.ACTION_DUR_FAST && group.contains(card)) {
            card.triggerOnExhaust()
            card.triggerOnManualDiscard()

            // TODO make a card sealing vfx -- see SoundMaster "KEY_OBTAIN"
            AbstractDungeon.effectList.add(ExhaustCardEffect(card))
            card.exhaustOnUseOnce = false
            card.freeToPlayOnce = false
            group.removeCard(card)
            AccursedMod.sealedPile.addToBottom(card)
            AbstractDungeon.player.onCardDrawOrDiscard()
        }
    }
}